File: WhatsApp-agents/waha/src/services/demoFlow.ts
//Mariano Montini ('bosque', 'bosquestudio')
import type { OutboundMessage } from '../contracts/outbound.js'
import {
  buildInteractiveButtonsMessage,
  buildInteractiveListMessage,
  buildMediaMessage,
  buildTextMessage,
} from './outboundBuilders.js'

// Demo keywords - inbound text triggers that open the menu or send a coupon.
const MENU_KEYWORDS = new Set(['hi', 'hello', 'hola', 'menu', 'start', 'ayuda'])
const COUPON_KEYWORDS = new Set(['coupon', 'cupon', 'pdf', 'promo', '2'])

// Interactive reply ids - deterministic button ids used in the demo menu.
export const DEMO_BUTTON_IDS = {
  info: 'menu_info',
  coupon: 'menu_coupon',
  help: 'menu_help',
} as const

// Demo flow options - coupon media URL and menu mode (text fallback vs native).
export type DemoFlowOptions = {
  couponMediaUrl: string
  menuMode?: 'text' | 'buttons' | 'list'
}

// Demo action - outbound message(s) the inbound handler should send.
export type DemoAction = {
  messages: OutboundMessage[]
}

// Resolve demo reply - maps keyword or interactive id to outbound Waha messages.
export function resolveDemoAction(
  input: { type: string; textBody?: string; interactiveId?: string },
  options: DemoFlowOptions,
): DemoAction {
  const to = '__TO__'
  const menuMode = options.menuMode ?? 'text'

  if (input.type === 'interactive' && input.interactiveId) {
    return resolveInteractive(input.interactiveId, to, options)
  }

  if (input.type !== 'text') {
    return { messages: [] }
  }

  const text = (input.textBody ?? '').trim().toLowerCase()

  if (text === '1' || text === 'info' || text === DEMO_BUTTON_IDS.info) {
    return resolveInteractive(DEMO_BUTTON_IDS.info, to, options)
  }
  if (text === '3' || text === 'help' || text === DEMO_BUTTON_IDS.help) {
    return resolveInteractive(DEMO_BUTTON_IDS.help, to, options)
  }

  if (MENU_KEYWORDS.has(text) || text === '') {
    return { messages: [buildMenuMessage(to, menuMode)] }
  }

  if (COUPON_KEYWORDS.has(text) || text === DEMO_BUTTON_IDS.coupon) {
    return {
      messages: [
        buildMediaMessage({
          to,
          type: 'document',
          link: options.couponMediaUrl,
          caption: 'Your demo coupon PDF',
          filename: 'coupon.pdf',
        }),
      ],
    }
  }

  return {
    messages: [
      buildTextMessage({
        to,
        body: `You said: "${input.textBody ?? ''}". Send "menu" for options or "coupon" for a PDF.`,
      }),
    ],
  }
}

// Build menu message - native buttons/list when requested, otherwise numbered text fallback.
function buildMenuMessage(to: string, menuMode: 'text' | 'buttons' | 'list'): OutboundMessage {
  const bodyText =
    'Choose an option to continue the Level 1 demo.\n1) Info\n2) Coupon PDF\n3) Help'

  if (menuMode === 'buttons') {
    return buildInteractiveButtonsMessage({
      to,
      headerText: 'Demo bot',
      bodyText,
      footerText: 'Waha template',
      buttons: [
        { id: DEMO_BUTTON_IDS.info, title: 'Info' },
        { id: DEMO_BUTTON_IDS.coupon, title: 'Coupon PDF' },
        { id: DEMO_BUTTON_IDS.help, title: 'Help' },
      ],
    })
  }

  if (menuMode === 'list') {
    return buildInteractiveListMessage({
      to,
      headerText: 'Demo bot',
      bodyText,
      buttonText: 'Options',
      footerText: 'Waha template',
      sections: [
        {
          title: 'Menu',
          rows: [
            { id: DEMO_BUTTON_IDS.info, title: 'Info' },
            { id: DEMO_BUTTON_IDS.coupon, title: 'Coupon PDF' },
            { id: DEMO_BUTTON_IDS.help, title: 'Help' },
          ],
        },
      ],
    })
  }

  return buildTextMessage({
    to,
    body: `Demo bot — ${bodyText}\nReply with 1, 2, 3, or keywords info / coupon / help.`,
  })
}

// Resolve interactive selection - maps button id to text or media follow-up.
function resolveInteractive(
  interactiveId: string,
  to: string,
  options: DemoFlowOptions,
): DemoAction {
  switch (interactiveId) {
    case DEMO_BUTTON_IDS.info:
      return {
        messages: [
          buildTextMessage({
            to,
            body: 'This is the Waha Level 1 demo: text, optional buttons/list (or text fallback), and media with presence+delay humanization.',
          }),
        ],
      }
    case DEMO_BUTTON_IDS.coupon:
      return {
        messages: [
          buildTextMessage({ to, body: 'Sending your coupon document…' }),
          buildMediaMessage({
            to,
            type: 'document',
            link: options.couponMediaUrl,
            caption: 'Demo coupon',
            filename: 'coupon.pdf',
          }),
        ],
      }
    case DEMO_BUTTON_IDS.help:
      return {
        messages: [
          buildTextMessage({
            to,
            body: 'Keywords: menu | coupon | hi. Options: 1 Info, 2 Coupon PDF, 3 Help. Native buttons/list only if WAHA_MENU_MODE is set and the engine supports them.',
          }),
        ],
      }
    default:
      return {
        messages: [
          buildTextMessage({
            to,
            body: `Unknown option "${interactiveId}". Send "menu" to try again.`,
          }),
        ],
      }
  }
}

// Bind recipient - replaces placeholder `to` with the real WhatsApp user id.
export function bindRecipient(messages: OutboundMessage[], to: string): OutboundMessage[] {
  return messages.map((message) => ({ ...message, to }))
}

File: WhatsApp-agents/whatsmeow/internal/services/demo_flow.go
package services

import "whatsmeow-agent/internal/contracts"

const (
	DemoButtonInfo   = "menu_info"
	DemoButtonCoupon = "menu_coupon"
	DemoButtonHelp   = "menu_help"
)

type DemoFlowOptions struct {
	CouponMediaURL string
	MenuMode       string
}

type DemoAction struct {
	Messages []contracts.OutboundMessage
}

// ResolveDemoAction - maps keyword or interactive id to outbound messages.
func ResolveDemoAction(inputType, textBody, interactiveID string, options DemoFlowOptions) DemoAction {
	to := "__TO__"
	menuMode := options.MenuMode
	if menuMode == "" {
		menuMode = "text"
	}

	if inputType == "interactive" && interactiveID != "" {
		return resolveInteractive(interactiveID, to, options)
	}
	if inputType != "text" {
		return DemoAction{}
	}

	text := stringsToLowerTrim(textBody)

	if text == "1" || text == "info" || text == DemoButtonInfo {
		return resolveInteractive(DemoButtonInfo, to, options)
	}
	if text == "3" || text == "help" || text == DemoButtonHelp {
		return resolveInteractive(DemoButtonHelp, to, options)
	}

	if isMenuKeyword(text) || text == "" {
		return DemoAction{Messages: []contracts.OutboundMessage{buildMenuMessage(to, menuMode)}}
	}

	if isCouponKeyword(text) || text == DemoButtonCoupon {
		return DemoAction{Messages: []contracts.OutboundMessage{BuildMediaMessage(contracts.MediaOutboundInput{
			To:       to,
			Type:     "document",
			Link:     options.CouponMediaURL,
			Caption:  "Your demo coupon PDF",
			Filename: "coupon.pdf",
		})}}
	}

	return DemoAction{Messages: []contracts.OutboundMessage{BuildTextMessage(contracts.TextOutboundInput{
		To:   to,
		Body: `You said: "` + textBody + `". Send "menu" for options or "coupon" for a PDF.`,
	})}}
}

// BindRecipient - replaces placeholder `to` with the real WhatsApp user id.
func BindRecipient(messages []contracts.OutboundMessage, to string) []contracts.OutboundMessage {
	out := make([]contracts.OutboundMessage, len(messages))
	for i, message := range messages {
		message.To = to
		out[i] = message
	}
	return out
}

func buildMenuMessage(to, menuMode string) contracts.OutboundMessage {
	bodyText := "Choose an option to continue the Level 1 demo.\n1) Info\n2) Coupon PDF\n3) Help"

	if menuMode == "buttons" {
		return BuildInteractiveButtonsMessage(to, bodyText)
	}
	if menuMode == "list" {
		return BuildInteractiveListMessage(to, bodyText)
	}

	return BuildTextMessage(contracts.TextOutboundInput{
		To:   to,
		Body: "Demo bot — " + bodyText + "\nReply with 1, 2, 3, or keywords info / coupon / help.",
	})
}

func resolveInteractive(interactiveID, to string, options DemoFlowOptions) DemoAction {
	switch interactiveID {
	case DemoButtonInfo:
		return DemoAction{Messages: []contracts.OutboundMessage{BuildTextMessage(contracts.TextOutboundInput{
			To:   to,
			Body: "This is the WhatsMeow Level 1 demo: text, numbered menu (native buttons unstable), and media with presence+delay humanization.",
		})}}
	case DemoButtonCoupon:
		return DemoAction{Messages: []contracts.OutboundMessage{
			BuildTextMessage(contracts.TextOutboundInput{To: to, Body: "Sending your coupon document…"}),
			BuildMediaMessage(contracts.MediaOutboundInput{
				To: to, Type: "document", Link: options.CouponMediaURL, Caption: "Demo coupon", Filename: "coupon.pdf",
			}),
		}}
	case DemoButtonHelp:
		return DemoAction{Messages: []contracts.OutboundMessage{BuildTextMessage(contracts.TextOutboundInput{
			To:   to,
			Body: "Keywords: menu | coupon | hi. Options: 1 Info, 2 Coupon PDF, 3 Help. Native buttons/list are unstable on WhatsMeow; default is numbered text.",
		})}}
	default:
		return DemoAction{Messages: []contracts.OutboundMessage{BuildTextMessage(contracts.TextOutboundInput{
			To:   to,
			Body: `Unknown option "` + interactiveID + `". Send "menu" to try again.`,
		})}}
	}
}

func isMenuKeyword(text string) bool {
	switch text {
	case "hi", "hello", "hola", "menu", "start", "ayuda":
		return true
	default:
		return false
	}
}

func isCouponKeyword(text string) bool {
	switch text {
	case "coupon", "cupon", "pdf", "promo", "2":
		return true
	default:
		return false
	}
}

func stringsToLowerTrim(value string) string {
	out := make([]rune, 0, len(value))
	started := false
	for _, r := range value {
		if !started && (r == ' ' || r == '\t' || r == '\n' || r == '\r') {
			continue
		}
		started = true
		if r >= 'A' && r <= 'Z' {
			r += 'a' - 'A'
		}
		out = append(out, r)
	}
	for len(out) > 0 {
		last := out[len(out)-1]
		if last == ' ' || last == '\t' || last == '\n' || last == '\r' {
			out = out[:len(out)-1]
			continue
		}
		break
	}
	return string(out)
}
